lib/CodeGen/CGDynamicCheck.cpp

```diff
@@ -111,6 +111,76 @@ void CodeGenFunction::EmitDynamicBoundsCheck(const Address PtrAddr, const Bounds
   EmitDynamicCheckBlocks(Builder.CreateAnd(LowerChk, UpperChk, "_Dynamic_check.range"));
 }
 
+void CodeGenFunction::EmitDynamicBoundsCheck(const Address BaseAddr,
+                                             const BoundsExpr *CastBounds,
+                                             const BoundsExpr *SubExprBounds) {
+  if (!getLangOpts().CheckedC)
+    return;
+
+  if (!SubExprBounds || !CastBounds)
+    return;
+
+  if (SubExprBounds->isAny() || SubExprBounds->isInvalid())
+    return;
+
+  // SubExprBounds can be Any by inference but CastBounds can't be Any
+  assert(!CastBounds->isAny());
+  if (CastBounds->isInvalid())
+    return;
+
+  // We can only generate the check if we have the bounds as a range.
+  if (!isa<RangeBoundsExpr>(SubExprBounds) ||
+      !isa<RangeBoundsExpr>(CastBounds)) {
+    llvm_unreachable(
+        "Can Only Emit Dynamic Bounds Check For RangeBounds Exprs");
+    return;
+  }
+
+  const RangeBoundsExpr *SubRange = dyn_cast<RangeBoundsExpr>(SubExprBounds);
+  const RangeBoundsExpr *CastRange = dyn_cast<RangeBoundsExpr>(CastBounds);
+
+  ++NumDynamicChecksRange;
+
+  // SubRange - bounds(lb, ub) vs CastRange - bounds(castlb, castub)
+  // Dynamic_check(lb <= castlb && castub <= ub)
+
+  // Emit the code to generate the pointer values
+  Address Lower = EmitPointerWithAlignment(SubRange->getLowerExpr());
+  Address Upper = EmitPointerWithAlignment(SubRange->getUpperExpr());
+
+  Value *LowerInt = Builder.CreatePtrToInt(Lower.getPointer(), IntPtrTy,
+                                           "_Dynamic_check.lower");
+  Value *UpperInt = Builder.CreatePtrToInt(Upper.getPointer(), IntPtrTy,
+                                           "_Dynamic_check.upper");
+
+  Address CastLower = EmitPointerWithAlignment(CastRange->getLowerExpr());
+  Address CastUpper = EmitPointerWithAlignment(CastRange->getUpperExpr());
+
+  Value *CastLowerInt = Builder.CreatePtrToInt(CastLower.getPointer(), IntPtrTy,
+                                               "_Dynamic_check.castlower");
+  Value *CastUpperInt = Builder.CreatePtrToInt(CastUpper.getPointer(), IntPtrTy,
+                                               "_Dynamic_check.castupper");
+
+  // Make the lower check (Lower <= CastLower)
+  Value *LowerChk =
+      Builder.CreateICmpULE(LowerInt, CastLowerInt, "_Dynamic_check.lower_cmp");
+
+  // Make the upper check (CastUpper <= Upper)
+  Value *UpperChk =
+      Builder.CreateICmpULE(CastUpperInt, UpperInt, "_Dynamic_check.upper_cmp");
+
+  // if expression E is NULL, skip dynamic_check
+  // Dynamic_check(E == NULL || (lb <= castlb && castub <= ub))
+  // if (cond1) {...}
+  // else {
+  // if (cond2) {...}
+  // else { trap(); llvm_unreachable(); }
+  // }
+  EmitDynamicCheckBlocks(
+      Builder.CreateIsNull(BaseAddr.getPointer(), "_Dynamic_check.null"),
```

Comment:
A newly implemented `EmitDynamicCheckBlocks(cond1, cond2)` generates code as follows:
```
if (cond1) { success_block }
else {
fallthrough_block
  if (cond2) {  success_block  }
  else {  fail_block  }
}
```
In current implementation, it emits dynamic check blocks for `(cond1|| cond2)`
cond1 is (E == NULL) and cond2 is (lb <= castlb && castub <= ub)
Therefore, if subexpression itself is NULL, it skips lower/upper bounds check
if subexpression is not NULL, then it also checks if lower/upper bounds condition is met or not
I think that it is equivalent for `dynamic_check (base == NULL || (lb <= castlb && castub <= ub))`